alc/backends/jack.cpp

```diff
@@ -503,11 +500,29 @@ bool JackBackendFactory::querySupport(BackendType type)
 std::string JackBackendFactory::probe(BackendType type)
 {
     std::string outnames;
+
+    auto list_ports = [&outnames] () -> void {
+     jack_client_t *client{nullptr};
+        jack_status_t status;
+        client = jack_client_open("alsoft", JackNoStartServer, &status, nullptr);
+        if(client == nullptr)
+            throw al::backend_exception{ALC_INVALID_VALUE, "Failed to open client connection: 0x%02x",
+                    status};
+        const char **ports{jack_get_ports(client, nullptr, nullptr, JackPortIsInput)};
+        for (int i = 0; ports && ports[i] ; i++) {
+            /* +1 to also append the null char (to ensure a null-separated list and
+            * double-null terminated list).
+            */
+            outnames.append(ports[i], strlen(ports[i]) + 1);
+        }
```

Comment:
Won't this report each port/channel as its own device? On my system, this gives `system:playback_1`, `system:playback_2`, etc, for each playback channel.

Reply:
Yes that's intended (for now).
It's patially flawed because I don't see a perfect solution to recover easyly the real outputs.
That's one of the 4 options I see to implement JackBackendFactory::probe (I implemented for now this one, because it's simple to implement and at least gives more possibility to to the user).
However, I don't know wich one would be the best.
- Return just "system" : this would break nothing, but still give the user the possibility to change manualy the output (ex : by changing the config file in OpenMW)
- Return all ports (as it is) : certainly not the best, simple to implement but will certainly break things
- Return JACK client names : give the user to choose beetween client freely (physicals output and simple clients will work). But some clients, like Ardour, will not work. But it will give the possibility to the user to choose "system" if nothing works.
- Ruturn the recovered real ouputs names : with the list of all ports, by cutting the end of the names by reconnizing some patterns (L/R, right/left, 1/2 or alternativly cut the last word after a '_' or a '-'). It will work on most clients and with jack's physical outputs, but may not work with some clients. I think this one is the best but it may confuses the user if he want to understand why their client doesn't work.

Which one do you think would be the best ?

Reply:
> It's patially flawed because I don't see a perfect solution to recover easyly the real outputs.

The main problem I see is that JACK doesn't really distinguish outputs. It just presents ports that each represent a playback channel to whatever devices are on the system (or act as a source to something else). There's no introspection to discover JACK devices (real or virtual), with the ports that belong to those specific devices. Consequently, the user is responsible for setting anything up.

One thought I had was looking through the ports and taking everything up to `:` as the device name (e.g. `system` for one device), although I don't know if that's a guaranteed pattern, or if different devices even have a different name before the colon. Cutting off the last word after ` `, `_`, or `-` and using the remaining as a name/pattern also feels sketchy and error-prone.

I think the more prudent option would be to allow the user to configure their own set of device names and port patterns in the config file. Something like:
```
[jack]
custom-devices = My Device=some_ports*;Another Device=other_ports*
```
Then OpenAL Soft would enumerate `JACK Default`, `My Device`, and `Another Device`, which would use the port name filters ``, `some_ports*`, and `other_ports*` respectively.

Reply:
>  although I don't know if that's a guaranteed pattern

Jack seems to delimit the device name and the "short" port name by the first `:`. So no problem with that.
[https://github.com/jackaudio/jack2/blob/98c882c5a545c3dd819655c9869f6928274df8a3/common/JackPort.cpp#L224](url)

>  I think the more prudent option would be to allow the user to configure their own set of device names and port patterns in the config file.

In this case, I think `<blank>` should list device names (`system`, `alsa_out`, ...) corresponding to physical outputs. So  every users may connect simply to the output of their choice (not all devices, because it may confuse unexperienced users if it doesn't work). Then if an advanced user want to use a non physical ports, they may use custom patterns like you said.